src/couch_replicator_httpc_pool.erl

```diff
@@ -172,6 +174,56 @@ terminate(_Reason, State) ->
     lists:foreach(fun ibrowse_http_client:stop/1, State#state.free),
     lists:foreach(fun ibrowse_http_client:stop/1, State#state.busy).
 
+
+release_worker_int(Worker, State0) ->
+    #state{
+        callers = Callers,
+        busy = Busy,
+        closing = Closing
+    } = State0,
+    State0#state{
+        callers = demonitor_client(Callers, Worker),
+        closing = Closing -- [Worker],
+        busy = Busy -- [Worker]
+    }.
+
+
+maybe_supply_worker(State) ->
+    #state{
+        url = Url,
+        waiting = Waiting0,
+        callers = Callers,
+        busy = Busy,
+        free = Free0,
+        closing = Closing,
+        limit = Limit
+    } = State,
+    case queue:out(Waiting0) of
+        {empty, Waiting0} ->
+            State;
+        {{value, From}, Waiting1} ->
+            case length(Busy) + length(Closing) >= Limit of
+                true ->
+                    State;
+                false ->
+                    {Worker, Free2} = case Free0 of
+                        [] ->
+                            {ok, W} = ibrowse:spawn_link_worker_process(Url),
+                            {W, Free0};
+                        [W|Free1] ->
+                            {W, Free1}
+                    end,
+                    gen_server:reply(From, {ok, Worker}),
+                    State#state{
+                        callers = monitor_client(Callers, Worker, From),
+                        waiting = Waiting1,
+                        busy = [Worker|Busy],
+                        free = Free2
+                    }
+            end
+    end.
+
+
 monitor_client(Callers, Worker, {ClientPid, _}) ->
```

Comment:
While you're at it, monitor_client is badly named or made, the Callers parameter and return value of a list of callers with only the first now monitored (the others already are) and added to the list, that seems kludgy to me.